src/metabase/lib/filter.cljc

```diff
@@ -26,16 +31,53 @@
   [query stage-number f]
   (->filter-arg query stage-number (f query stage-number)))
 
-(mu/defn = :- [:or
-               fn?
-               :mbql.clause/=]
-  "Create an `=` filter clause."
-  ([x y]
-   (fn [query stage-number]
-     (= query stage-number x y)))
-  ([query stage-number x y & more]
-   (-> (into [:=]
-             (map (fn [arg]
-                    (->filter-arg query stage-number arg)))
-             (list* x y more))
-       lib.options/ensure-uuid)))
+#?(:clj
+   (defmacro ^:private deffilter
+     [filter-name argvec]
+     {:pre [(symbol? filter-name)
+            (vector? argvec) (every? symbol? argvec)
+            (not-any? #{'query 'stage-number} argvec)]}
+     (let [filter-name-str (name filter-name)
+           vararg? (.contains argvec '&)
+           args (remove #{'&} argvec)
+           arglist-expr (if vararg?
+                          (cons 'list* args)
+                          argvec)]
+       `(do
+          (mu/defn ~filter-name :- ~(keyword "mbql.clause" filter-name-str)
+            ~(format "Create a filter clause of type `%s`." filter-name-str)
+            [~'query ~'stage-number ~@argvec]
+            (-> (into [~(keyword filter-name)]
+                      (map (fn [~'arg]
+                             (->filter-arg ~'query ~'stage-number ~'arg)))
+                      ~arglist-expr)
+                metabase.lib.options/ensure-uuid))
+
+          (mu/defn ~(symbol (str "->" filter-name-str)) :- fn?
+            ~(format "Return function creating a filter clause of type `%s`." filter-name-str)
+            ~argvec
+            (fn [~'query ~'stage-number]
+              ~(cond->> (concat [filter-name 'query 'stage-number] args)
+                 vararg? (cons `apply))))))))
```

Comment:
- I prefer not to mix the filter and the filter thunk creator functions under a single name, because I think they are used in different situations. (The "normal" and transducer arities of `map` is a counterexample, so I might be wrong to feel this way.)
- Just simplistically putting the two functions under a single name is somewhat tricky, because we cannot have more than one vararg arity and whole thing can get ambiguous.
- The `->` prefix is not the best, e.g. `->>` is difficult to associate with the `>` thunk creator. Suggestions are welcome!